x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/InferenceQueryUtils.java

```diff
@@ -0,0 +1,677 @@
+            // - The query is null
+            // We perform local inference result generation if an inference ID override is specified and the query is non-null because
+            // remote cluster fields (either inference or non-inference) may need this inference result to handle the query.
+            localInferenceInfoListener.onResponse(
+                new InferenceInfo(
+                    inferenceFieldCount,
+                    indexCount,
+                    inferenceResultsMap != null ? inferenceResultsMap : Map.of(),
+                    queryRewriteContext.getMinTransportVersion()
+                )
+            );
+            return;
+        }
+
+        Set<FullyQualifiedInferenceId> localInferenceIds;
+        boolean useCoordinatedInferenceAction;
+        if (inferenceIdOverride != null) {
+            // Use CoordinatedInferenceAction when an override is set because the override could refer to trained model
+            localInferenceIds = Set.of(inferenceIdOverride);
+            useCoordinatedInferenceAction = true;
+        } else {
+            localInferenceIds = getLocalInferenceIds(localInferenceFields, queryRewriteContext.getLocalClusterAlias());
+            useCoordinatedInferenceAction = false;
+        }
+
+        final int finalInferenceFieldCount = inferenceFieldCount;
+        getLocalInferenceResults(
+            queryRewriteContext,
+            query,
+            localInferenceIds,
+            inferenceResultsMap,
+            useCoordinatedInferenceAction,
+            localInferenceInfoListener.delegateFailureAndWrap((l, m) -> {
+                InferenceInfo inferenceInfo = new InferenceInfo(
+                    finalInferenceFieldCount,
+                    indexCount,
+                    m,
+                    queryRewriteContext.getMinTransportVersion()
+                );
+                l.onResponse(inferenceInfo);
+            })
+        );
+    }
+
+    private static void getRemoteInferenceInfo(
+        QueryRewriteContext queryRewriteContext,
+        Map<String, Float> fields,
+        boolean resolveWildcards,
+        boolean useDefaultFields,
+        ActionListener<Map<String, Tuple<GetInferenceFieldsAction.Response, TransportVersion>>> remoteInferenceInfoListener,
+        @Nullable String query,
+        @Nullable FullyQualifiedInferenceId inferenceIdOverride
+    ) {
+        var remoteIndices = queryRewriteContext.getResolvedIndices().getRemoteClusterIndices();
+        GroupedActionListener<Map<String, Tuple<GetInferenceFieldsAction.Response, TransportVersion>>> gal =
+            createRemoteInferenceInfoGroupedActionListener(remoteIndices.size(), remoteInferenceInfoListener);
+
+        // When an inference ID override is set, inference is only performed on the local cluster. Set the query to null in this case
+        // to disable remote inference result generation.
+        String effectiveQuery = inferenceIdOverride == null ? query : null;
```

Comment:
This might be a good one to make `final`

Reply:
++